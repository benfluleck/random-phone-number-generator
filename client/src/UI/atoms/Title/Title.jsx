import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import setClassNames from '../../../helpers/utils/setClassNames';

const Title = ({ children }) => (
  <Title.Container className={setClassNames({ 'is-size-1': true, 'has-text-centered': true })}>
    {children}
  </Title.Container>
);

export default Title;

Title.Container = styled.h2``;

Title.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
};
