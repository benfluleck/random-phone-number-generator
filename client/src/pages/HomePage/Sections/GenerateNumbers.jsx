import React from 'react'
import styled from 'styled-components';

import setClassNames from '../../../utils/setClassNames'

const GenerateNumbers = ({ title, numbers, handleOnChange, isDisabled, errorMessages }) => (
  <GenerateNumbers.Container className="section">
    <label className={
      setClassNames({
        'is-size-5': true,
        'has-text-centered': true,
        'has-text-weight-light': true,
        'label': true
      })}>
      {title} &nbsp;
    </label>
    <input
      className={
        setClassNames({
          "input": true,
          "is-medium": true,
          "is-danger": errorMessages
        })}
      type="text"
      id="addInput"
      value={numbers}
      placeholder="Number of Phone Numbers"
      onChange={handleOnChange}
      disabled={isDisabled && 'disabled'} />
    <p className="is-size-6 has-text-danger has-text-centered has-text-weight-light label">{errorMessages}</p>
  </GenerateNumbers.Container>)


GenerateNumbers.Container = styled.section`
`;


export default GenerateNumbers
