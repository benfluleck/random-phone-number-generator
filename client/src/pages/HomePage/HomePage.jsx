import React, { Component, Fragment } from 'react';

import validator from '../../helpers/utils/validator';
import axiosConfig from '../../config/axiosConfig';
import Button from '../../UI/atoms/Button/Button';
import Title from '../../UI/atoms/Title/Title';
import GenerateNumbers from './Sections/GenerateNumbers';

class HomePage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      numberOfPhonenumbers: '',
      errorMessage: {},
      isDisabled: false,
      isChecked: false,
    };
    this.handleOnChange = this.handleOnChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.toggleCheckbox = this.toggleCheckbox.bind(this);
  }

  handleOnChange(e) {
    this.setState({
      errorMessage: { numberOfPhonenumbers: '' },
    });
    this.setState({ numberOfPhonenumbers: e.target.value });
  }

  toggleCheckbox() {
    const { isChecked } = this.state;

    this.setState({ isChecked: !isChecked });
  }

  handleClick(e) {
    e.preventDefault();

    const dataRules = {
      numberOfPhonenumbers: [ 'upperLimit', 'lowerLimit' ],
    };
    const messages = {
      numberOfPhonenumbers: {
        upperLimit:
          'You can not download more than 10,000 numbers',
        lowerLimit: 'Your number needs to be higher than 1',
      },
    };

    const {
      numberOfPhonenumbers,
      isChecked,
    } = this.state;

    const validatorResult = validator({ numberOfPhonenumbers }, dataRules, messages);

    if (!validatorResult.isValid) {
      this.setState({ errorMessage: validatorResult.errors });
    } else {
      let query;

      isChecked ? query = 'desc' : query = '';

      this.setState({ isDisabled: true }, () => {
        axiosConfig
          .phoneNumbers
          .generate({ numberOfPhonenumbers }, query)
          .then(() => {
            this.props.history.push('/results');
          });
      });
    }
  }

  render() {
    const {
      numberOfPhonenumbers,
      isDisabled,
      errorMessage: { numberOfPhonenumbers: errors }
    } = this.state;

    return (
      <Fragment>
        <Title>
          Phone Number Generator
        </Title>
        <GenerateNumbers
          numbers={numberOfPhonenumbers}
          title="Please Enter Number of Phone Numbers"
          handleOnChange={this.handleOnChange}
          isDisabled={isDisabled}
          errorMessages={errors}
        />
        <label className="checkbox has-text-grey-dark is-size-6">
          <input type="checkbox" onClick={this.toggleCheckbox} />
             &nbsp;Descending Order: Default file arrangement is Ascending
        </label>
        <p className="is-size-6 help" />
        <Button
          isDisabled={isDisabled}
          onClick={this.handleClick}>
            Generate Numbers
        </Button>
      </Fragment>
    );
  }
}

export default HomePage;
