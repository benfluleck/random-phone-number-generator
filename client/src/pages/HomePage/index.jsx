import React, { Component, Fragment } from 'react';

import validator from '../../utils/validator';
import axiosConfig from '../../config/axiosConfig';
import Button from '../../UI/atoms/Button/Button'
import Title from '../../UI/atoms/Title/Title'
import GenerateNumbers from './Sections/GenerateNumbers';

class HomePage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      numberOfPhonenumbers: '',
      errorMessage: {},
      isDisabled: false
    };
    this.handleOnChange = this.handleOnChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  handleOnChange(e) {
    this.setState({
      errorMessage: { numberOfPhonenumbers: '' },
    });
    this.setState({ numberOfPhonenumbers: e.target.value });
  }

  handleClick(e) {
    e.preventDefault();

    const dataRules = {
      numberOfPhonenumbers: ['upperLimit', 'lowerLimit'],
    };
    const messages = {
      numberOfPhonenumbers: {
        upperLimit:
          'You can not download more than 10,000 numbers',
        lowerLimit: 'Your number needs to be higher than 1',
      },
    };

    const {
      numberOfPhonenumbers
    } = this.state;

    const validatorResult = validator({ numberOfPhonenumbers }, dataRules, messages);

    if (!validatorResult.isValid) {
      this.setState({ errorMessage: validatorResult.errors });
    } else {
      this.setState({ isDisabled: true }, () => {

        axiosConfig
          .phoneNumbers
          .generate({ numberOfPhonenumbers })
          .then((response) => {
            console.log('I have returned something', response);
          });
      })
    }
  }

  render() {
    const {
      numberOfPhonenumbers,
      isDisabled,
      errorMessage: { numberOfPhonenumbers: errors }
    } = this.state;

    return (
      <Fragment>
        <Title>
          Phone Number Generator
        </Title>
        <GenerateNumbers
          numbers={numberOfPhonenumbers}
          title="Please Enter Number of Phone Numbers"
          handleOnChange={this.handleOnChange}
          isDisabled={isDisabled}
          errorMessages={errors}
          />
        <Button
          isDisabled={isDisabled}
          onClick={this.handleClick}>
          Generate Numbers
        </Button>
      </Fragment>
    );
  }
}



export default HomePage;
