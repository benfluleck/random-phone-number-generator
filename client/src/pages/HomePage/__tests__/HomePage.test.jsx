import React from 'react';
import { shallow, mount } from 'enzyme';

import HomePage from '../HomePage';
import Title from '../../../UI/atoms/Title/Title';
import GenerateNumbers from '../Sections/GenerateNumbers';


const historyMock = { push: jest.fn() };

const setup = (props) => {
  const wrapper = shallow(
    <HomePage {...props} />
  );

  const mountedWrapper = mount(
    <HomePage {...props} />
  );

  return {
    wrapper,
    mountedWrapper,
  };
};


describe('Component - Home Page', () => {
  test('should render', () => {
    const { wrapper } = setup();

    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
  test('should render a title', () => {
    const { wrapper } = setup();

    expect(wrapper.contains(<Title>Phone Number Generator</Title>)).toBeTruthy();
  });
  test('should render GeneratedNumber Component with props', () => {
    const { wrapper } = setup();

    const generatedWrapper = wrapper.setState({ numberOfPhonenumbers: '23', errorMessage: 'This is an error' });

    expect(generatedWrapper.find(GenerateNumbers).props().numbers).toEqual('23');
    expect(generatedWrapper.find(GenerateNumbers).props().title).toEqual('Please Enter Number of Phone Numbers');
  });
  test('should toggle Checkbox click', () => {
    const { mountedWrapper } = setup();

    const handletoggleCheckboxSpy = jest.spyOn(
      mountedWrapper.instance(),
      'toggleCheckbox'
    );

    mountedWrapper.instance().toggleCheckbox();
    expect(handletoggleCheckboxSpy).toHaveBeenCalledTimes(1);
  });

  test('should simulate onchange if input is changed', () => {
    const { mountedWrapper } = setup();

    mountedWrapper.find('input').first().simulate('change', { target: { value: '24' } });

    expect(mountedWrapper.state('numberOfPhonenumbers')).toEqual('24');
  });

  test('should simulate click button and return error message with no input', () => {
    const { mountedWrapper } = setup({ history: historyMock });

    const mockEvent = { preventDefault: jest.fn() };

    mountedWrapper.find('button').simulate('click', mockEvent);

    expect(mountedWrapper.state('errorMessage').numberOfPhonenumbers).toEqual('Your number needs to be higher than 1');
  });
  test('should simulate click button is clicked', () => {
    const { mountedWrapper } = setup({ history: historyMock });

    const mockEvent = { preventDefault: jest.fn() };

    mountedWrapper.setState({ isChecked: true, numberOfPhonenumbers: '45' });
    mountedWrapper.find('button').simulate('click', mockEvent);

    expect(mountedWrapper.state('errorMessage')).toEqual({});
  });
});
