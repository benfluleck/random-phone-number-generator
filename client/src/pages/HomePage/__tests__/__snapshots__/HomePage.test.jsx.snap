// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Home Page should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HomePage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Title>
          Phone Number Generator
        </Title>,
        <GenerateNumbers
          errorMessages=""
          handleOnChange={[Function]}
          isDisabled={false}
          numbers=""
          title="Please Enter Number of Phone Numbers"
        />,
        <label
          className="checkbox has-text-grey-dark is-size-6"
        >
          <input
            onClick={[Function]}
            type="checkbox"
          />
           Descending Order: Default file arrangement is Ascending
        </label>,
        <p
          className="is-size-6 help"
        />,
        <Button
          isDisabled={false}
          left={false}
          onClick={[Function]}
          right={false}
        >
          Generate Numbers
        </Button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Phone Number Generator",
        },
        "ref": null,
        "rendered": "Phone Number Generator",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "errorMessages": "",
          "handleOnChange": [Function],
          "isDisabled": false,
          "numbers": "",
          "title": "Please Enter Number of Phone Numbers",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onClick={[Function]}
              type="checkbox"
            />,
            " Descending Order: Default file arrangement is Ascending",
          ],
          "className": "checkbox has-text-grey-dark is-size-6",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onClick": [Function],
              "type": "checkbox",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          " Descending Order: Default file arrangement is Ascending",
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "is-size-6 help",
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Generate Numbers",
          "isDisabled": false,
          "left": false,
          "onClick": [Function],
          "right": false,
        },
        "ref": null,
        "rendered": "Generate Numbers",
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Title>
            Phone Number Generator
          </Title>,
          <GenerateNumbers
            errorMessages=""
            handleOnChange={[Function]}
            isDisabled={false}
            numbers=""
            title="Please Enter Number of Phone Numbers"
          />,
          <label
            className="checkbox has-text-grey-dark is-size-6"
          >
            <input
              onClick={[Function]}
              type="checkbox"
            />
             Descending Order: Default file arrangement is Ascending
          </label>,
          <p
            className="is-size-6 help"
          />,
          <Button
            isDisabled={false}
            left={false}
            onClick={[Function]}
            right={false}
          >
            Generate Numbers
          </Button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Phone Number Generator",
          },
          "ref": null,
          "rendered": "Phone Number Generator",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "errorMessages": "",
            "handleOnChange": [Function],
            "isDisabled": false,
            "numbers": "",
            "title": "Please Enter Number of Phone Numbers",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onClick={[Function]}
                type="checkbox"
              />,
              " Descending Order: Default file arrangement is Ascending",
            ],
            "className": "checkbox has-text-grey-dark is-size-6",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onClick": [Function],
                "type": "checkbox",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            " Descending Order: Default file arrangement is Ascending",
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "is-size-6 help",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Generate Numbers",
            "isDisabled": false,
            "left": false,
            "onClick": [Function],
            "right": false,
          },
          "ref": null,
          "rendered": "Generate Numbers",
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
